File: rust/src/bin/aoc_2022_14.rs
// Advent of Code, 2022 day 14

use aoc_rust::stdin_lines;
use aoc_rust::UCoord;
use grid::Grid;
use std::iter::zip;

// We add sand in two modes
#[derive(Clone, Copy, PartialEq)]

enum AddSandMode {
    UntilOverflow,
    UntilReachTop,
}

#[derive(Clone, Copy, Default, PartialEq)]
enum Square {
    #[default]
    Empty,
    Sand,
    Rock,
}

// Start point must be at row 0, col START_COL
const START_COL: usize = 500;

struct Board {
    rock_row_max: usize, // lowest non-floor rock
    row_max: usize,      // floor row
    col_min: usize,
    cols: usize,
    finished: bool,
    grid: Grid<Square>,
}

impl Board {
    fn new(lines: &[String]) -> Board {
        let paths: Vec<Vec<UCoord>> = lines.iter().map(|s| parse_line(s)).collect();

        // Find extent of rocks
        let mut rock_row_min: usize = 0;
        let mut rock_row_max: usize = 0;
        let mut rock_col_min: usize = START_COL;
        let mut rock_col_max: usize = START_COL;
        for path in &paths {
            for c in path {
                rock_row_min = rock_row_min.min(c.row);
                rock_row_max = rock_row_max.max(c.row);
                rock_col_min = rock_col_min.min(c.col);
                rock_col_max = rock_col_max.max(c.col);
            }
        }
        assert!(rock_row_min == 0);

        // Define extent of board grid
        let row_max: usize = rock_row_max + 2;
        let rows: usize = row_max - rock_row_min + 1;
        let col_delta: usize = rows + 1;
        let col_min: usize = rock_col_min - col_delta;
        let col_max: usize = rock_col_max + col_delta;
        let cols: usize = col_max - col_min + 1;

        // Create board
        let mut board = Board {
            rock_row_max,
            row_max,
            col_min,
            cols,
            finished: false,
            grid: Grid::new(rows, cols),
        };

        // Fill board grid with rocks
        for path in paths {
            for (c1, c2) in zip(path.iter(), path[1..].iter()) {
                board.add_rock_line(c1, c2);
            }
        }
        // Add rocks to the floor
        for c in col_min..=col_max {
            board.set(
                UCoord {
                    row: row_max,
                    col: c,
                },
                Square::Rock,
            );
        }
        board
    }

    fn add(&mut self, add_sand_mode: AddSandMode) {
        let mut c = UCoord {
            row: 0,
            col: START_COL,
        };
        assert!(self.get(c) == Square::Empty);
        loop {
            // for part (a), finish when overflow lowest rock
            if add_sand_mode == AddSandMode::UntilOverflow && c.row >= self.rock_row_max {
                self.finished = true;
                return;
            }
            // Check down
            let mut d = c + UCoord { row: 1, col: 0 };
            if self.get(d) == Square::Empty {
                c = d;
                continue;
            }
            // Check down-left
            d -= UCoord { row: 0, col: 1 };
            if self.get(d) == Square::Empty {
                c = d;
                continue;
            }
            // Check down-right
            d += UCoord { row: 0, col: 2 };
            if self.get(d) == Square::Empty {
                c = d;
                continue;
            }
            // Settle
            self.set(c, Square::Sand);
            // for part(b), finish when reach top
            if add_sand_mode == AddSandMode::UntilReachTop && c.row == 0 {
                self.finished = true;
            }
            break;
        }
    }

    fn add_until_overflow(&mut self, add_sand_mode: AddSandMode) -> usize {
        let mut sand_count: usize = 0;
        self.finished = false;
        while !self.finished {
            self.add(add_sand_mode);
            sand_count += 1;
        }
        sand_count
    }

    // Add rocks to all points between the two coordinates (inclusive)
    fn add_rock_line(&mut self, c1: &UCoord, c2: &UCoord) {
        if c2.col > c1.col {
            assert!(c2.row == c1.row);
            for c in c1.col..=c2.col {
                self.set(
                    UCoord {
                        row: c1.row,
                        col: c,
                    },
                    Square::Rock,
                );
            }
        } else if c2.col < c1.col {
            assert!(c2.row == c1.row);
            for c in c2.col..=c1.col {
                self.set(
                    UCoord {
                        row: c1.row,
                        col: c,
                    },
                    Square::Rock,
                );
            }
        } else if c2.row > c1.row {
            assert!(c2.col == c1.col);
            for r in c1.row..=c2.row {
                self.set(
                    UCoord {
                        row: r,
                        col: c1.col,
                    },
                    Square::Rock,
                );
            }
        } else {
            assert!(c2.row < c1.row);
            assert!(c2.col == c1.col);
            for r in c2.row..=c1.row {
                self.set(
                    UCoord {
                        row: r,
                        col: c1.col,
                    },
                    Square::Rock,
                );
            }
        }
    }

    fn get(&self, c: UCoord) -> Square {
        assert!(c.row <= self.row_max);
        assert!(c.col >= self.col_min);
        assert!(c.col < self.col_min + self.cols);
        self.grid[c.row][c.col - self.col_min]
    }

    fn set(&mut self, c: UCoord, value: Square) {
        assert!(c.row <= self.row_max);
        assert!(c.col >= self.col_min);
        assert!(c.col < self.col_min + self.cols);
        self.grid[c.row][c.col - self.col_min] = value;
    }
}

fn parse_coord(s: &str) -> UCoord {
    let mut iter = s.split(',');
    let word1 = iter.next().unwrap();
    let word2 = iter.next().unwrap();
    assert!(iter.next().is_none());
    UCoord {
        row: word2.parse().unwrap(),
        col: word1.parse().unwrap(),
    }
}

fn parse_line(s: &str) -> Vec<UCoord> {
    s.split(" -> ").map(parse_coord).collect()
}

fn main() {
    let lines: Vec<String> = stdin_lines().collect();
    let mut board = Board::new(&lines);
    let overflow_sand = board.add_until_overflow(AddSandMode::UntilOverflow) - 1;
    println!("{overflow_sand}");
    let extra_sand = board.add_until_overflow(AddSandMode::UntilReachTop);
    println!("{}", overflow_sand + extra_sand);
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_parse_coord() {
        assert_eq!(parse_coord("497,67"), UCoord { row: 67, col: 497 });
    }
}

File: others/zig/src/aoc_2022_14.zig
// Advent of Code, 2022 day 14

const std = @import("std");
const rw = @import("rw");
const Coord = @import("coord").Coord(i64);

const Square = enum { empty, sand, rock };

const Grid = struct {
    rock_row_max: i64, // lowest non-floor rock
    row_max: i64, // floor row
    rows: usize,
    col_min: i64,
    cols: usize,
    data: []Square,
    allocator: std.mem.Allocator,
    finished: bool,

    const start = Coord.init(0, 500);

    fn get(self: *Grid, c: Coord) Square {
        std.debug.assert(c.row >= Grid.start.row);
        std.debug.assert(c.col >= self.col_min);
        const row_diff: usize = @intCast(c.row - Grid.start.row);
        const col_diff: usize = @intCast(c.col - self.col_min);
        std.debug.assert(row_diff < self.rows);
        std.debug.assert(col_diff < self.cols);
        return self.data[row_diff * self.cols + col_diff];
    }

    fn set(self: *Grid, c: Coord, value: Square) void {
        const row_diff: usize = @intCast(c.row - Grid.start.row);
        const col_diff: usize = @intCast(c.col - self.col_min);
        self.data[row_diff * self.cols + col_diff] = value;
    }

    fn init(allocator: std.mem.Allocator, lines: [][]const u8) !Grid {

        // Convert text lines to paths
        var paths = std.ArrayList([]Coord).init(allocator);
        for (lines) |line| {
            try paths.append(try parse_line(allocator, line));
        }
        defer {
            for (paths.items) |path| {
                allocator.free(path);
            }
            paths.deinit();
        }

        // Find extent of board
        var rock_row_min: i64 = Grid.start.row;
        var rock_row_max: i64 = Grid.start.row;
        var rock_col_min: i64 = Grid.start.col;
        var rock_col_max: i64 = Grid.start.col;
        for (paths.items) |path| {
            for (path) |c| {
                rock_row_min = @min(rock_row_min, c.row);
                rock_row_max = @max(rock_row_max, c.row);
                rock_col_min = @min(rock_col_min, c.col);
                rock_col_max = @max(rock_col_max, c.col);
            }
        }
        std.debug.assert(rock_row_min == Grid.start.row);
        const row_max: i64 = rock_row_max + 2;
        const rows: usize = @intCast(row_max - rock_row_min + 1);
        const col_delta: i64 = @intCast(rows + 1);
        const col_min: i64 = rock_col_min - col_delta;
        const col_max: i64 = rock_col_max + col_delta;
        const cols: usize = @intCast(col_max - col_min + 1);

        // Create board
        var grid = Grid{
            .rock_row_max = rock_row_max,
            .row_max = row_max,
            .rows = rows,
            .col_min = col_min,
            .cols = cols,
            .data = try allocator.alloc(Square, rows * cols),
            .finished = false,
            .allocator = allocator,
        };

        // Fill board
        for (grid.data) |*d| {
            d.* = Square.empty;
        }
        for (paths.items) |path| {
            for (path[0 .. path.len - 1], path[1..path.len]) |c1, c2| {
                grid.add_rock_line(c1, c2);
            }
        }
        var c = col_min;
        while (c <= col_max) {
            grid.set(Coord.init(row_max, c), Square.rock);
            c += 1;
        }

        return grid;
    }

    fn deinit(self: *Grid) void {
        self.allocator.free(self.data);
    }

    fn add_rock_line(self: *Grid, a: Coord, b: Coord) void {
        const delta = b.sub(a).sign();
        var c = a;
        while (true) {
            self.set(c, Square.rock);
            if (c.eql(b)) {
                break;
            } else {
                c = c.add(delta);
            }
        }
    }

    // Add one unit of sand
    fn add(self: *Grid, floor: bool) void {
        var c = Grid.start;
        std.debug.assert(self.get(c) == Square.empty);
        while (true) {
            if (!floor and c.row >= self.rock_row_max) {
                self.finished = true;
                break;
            }
            var d = c.add(Coord.init(1, 0));
            if (self.get(d) == Square.empty) {
                c = d;
                continue;
            }
            d = c.add(Coord.init(1, -1));
            if (self.get(d) == Square.empty) {
                c = d;
                continue;
            }
            d = c.add(Coord.init(1, 1));
            if (self.get(d) == Square.empty) {
                c = d;
                continue;
            }
            self.set(c, Square.sand);
            if (floor and c.eql(Grid.start)) {
                self.finished = true;
                break;
            }
            break;
        }
    }
};

fn parse_coord(s: []const u8) !Coord {
    var iter = std.mem.splitScalar(u8, s, ',');
    const word1 = iter.next().?;
    const word2 = iter.next().?;
    std.debug.assert(iter.next() == null);
    return Coord{
        .row = try std.fmt.parseInt(i64, word2, 10),
        .col = try std.fmt.parseInt(i64, word1, 10),
    };
}

fn parse_line(allocator: std.mem.Allocator, s: []const u8) ![]Coord {
    var iter = std.mem.splitSequence(u8, s, " -> ");
    var list = std.ArrayList(Coord).init(allocator);
    while (iter.next()) |word| {
        try list.append(try parse_coord(word));
    }
    return list.toOwnedSlice();
}

fn add_until_overflow(grid: *Grid, floor: bool) usize {
    var sand_count: usize = 0;
    grid.finished = false;
    while (!grid.finished) {
        grid.add(floor);
        sand_count += 1;
    }
    return sand_count;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read text input lines
    const lines: [][]const u8 = try rw.read_lines(allocator);
    defer {
        for (lines) |line| {
            allocator.free(line);
        }
        allocator.free(lines);
    }

    var grid = try Grid.init(allocator, lines);
    defer grid.deinit();

    const overflow_sand: usize = add_until_overflow(&grid, false) - 1;
    try rw.write_int(usize, overflow_sand);
    const extra_sand: usize = add_until_overflow(&grid, true);
    try rw.write_int(usize, overflow_sand + extra_sand);
}
